sockeye/train.py

```diff
@@ -1008,7 +1001,18 @@ def train(args: argparse.Namespace, custom_metrics_logger: Optional[Callable] =
 
     utils.log_parameters(sockeye_model)
 
-    optimizer, zero_grad_kwargs = optimizers.get_optimizer(sockeye_model, optimizer_config)
+    losses = create_losses(args, all_num_classes=target_vocab_sizes)
+
+    optimizer_class, optimizer_kwargs, zero_grad_kwargs = optimizers.get_optimizer(optimizer_config)
+    optimizer = optimizer_class(sockeye_model.parameters(), **optimizer_kwargs)
+
+    lr_scheduler_class, lr_scheduler_kwargs = lr_scheduler.get_lr_scheduler(args.learning_rate_scheduler_type,
+                                                                            args.initial_learning_rate,
+                                                                            args.learning_rate_reduce_factor,
+                                                                            args.learning_rate_reduce_num_not_improved,
+                                                                            args.learning_rate_warmup,
+                                                                            args.max_updates)
+    _lr_scheduler = lr_scheduler_class(optimizer, **lr_scheduler_kwargs) if lr_scheduler_class is not None else None
```

Comment:
What's the reason to split this into the class and the kwargs rather than passing the optimizer to `get_lr_scheduler`?
(is this related to upcoming changes for DeepSpeed?)

Reply:
Yes. The idea is that our functions for creating the optimizer and learning rate scheduler return a class and the kwargs for instantiating that class. We can instantiate the class directly or pass the class/kwargs to DeepSpeed. For example, DeepSpeed matches the optimizer name and instantiates its own implementation (DeepSpeed Adam instead of PyTorch Adam) using the same kwargs: https://github.com/awslabs/sockeye/blob/deepspeed/sockeye/train.py#L822-L824